File: src/tools/review.ts
import { readFile } from "node:fs/promises";
import { z } from "zod";
import type { AgentDefinition } from "../types.js";
import { runLinters, type LinterRunResult } from "./linter.js";

const ReviewInputSchema = z.object({
  file_path: z.string().describe("Absolute path to the file to review"),
  context: z
    .string()
    .optional()
    .describe("Additional context about the file or review focus"),
});

export type ReviewInput = z.infer<typeof ReviewInputSchema>;

export { ReviewInputSchema };

export const reviewInputJsonSchema = {
  type: "object" as const,
  properties: {
    file_path: {
      type: "string" as const,
      description: "Absolute path to the file to review",
    },
    context: {
      type: "string" as const,
      description: "Additional context about the file or review focus",
    },
  },
  required: ["file_path"],
};

function formatLinterFindings(result: LinterRunResult): string {
  const { findings, skipped, errors } = result;

  // Nothing to report at all
  if (findings.length === 0 && skipped.length === 0 && errors.length === 0) {
    return "";
  }

  const lines: string[] = ["\n## Linter Findings\n"];

  if (findings.length > 0) {
    for (const f of findings) {
      const location = [f.file, f.line].filter(Boolean).join(":");
      const prefix = location ? `${location} — ` : "";
      lines.push(`**${f.tool}** [${f.severity}] ${prefix}${f.message}`);
    }
  } else {
    lines.push("No issues found by linters.");
  }

  if (skipped.length > 0) {
    lines.push("");
    lines.push(`**Not installed (skipped):** ${skipped.join(", ")}`);
  }

  if (errors.length > 0) {
    lines.push("");
    lines.push(`**Errors:** ${errors.join("; ")}`);
  }

  return lines.join("\n");
}

export function buildReviewPrompt(
  agent: AgentDefinition,
  fileContent: string,
  filePath: string,
  context?: string,
  linterOutput?: string
): string {
  const parts: string[] = [];

  parts.push(agent.system_prompt);

  if (agent.checklist) {
    parts.push("\n## Review Checklist\n");
    parts.push(agent.checklist);
  }

  parts.push("\n## File Under Review\n");
  parts.push(`**Path:** \`${filePath}\`\n`);

  // Add language hint from file extension
  const ext = filePath.split(".").pop() ?? "";
  const langMap: Record<string, string> = {
    ts: "typescript",
    tsx: "typescript",
    js: "javascript",
    jsx: "javascript",
    py: "python",
    rs: "rust",
    go: "go",
    java: "java",
    php: "php",
    cs: "csharp",
    c: "c",
    h: "c",
    dart: "dart",
    lua: "lua",
    sql: "sql",
    proto: "protobuf",
  };

  parts.push("```" + (langMap[ext] ?? ""));
  parts.push(fileContent);
  parts.push("```");

  if (linterOutput) {
    parts.push(linterOutput);
  }

  if (context) {
    parts.push("\n## Additional Context\n");
    parts.push(context);
  }

  // Linter cross-reference note (only when linter output is available)
  if (linterOutput) {
    parts.push("\n## Note\n");
    parts.push("If linter findings are provided above, reference them where relevant — confirm, expand on, or contextualize what the tools found.");
  }

  return parts.join("\n");
}

export async function executeReview(
  agent: AgentDefinition,
  input: ReviewInput
): Promise<{ content: string; isError: boolean }> {
  let fileContent: string;
  try {
    fileContent = await readFile(input.file_path, "utf-8");
  } catch (err) {
    return {
      content: `Cannot read file: ${input.file_path} — ${err instanceof Error ? err.message : String(err)}. Check that the path is absolute and the file exists.`,
      isError: true,
    };
  }

  // Run linters (best-effort — don't fail the review if linters break)
  let linterOutput = "";
  try {
    const linterResult = await runLinters(input.file_path, agent.recommended_tools);
    linterOutput = formatLinterFindings(linterResult);
  } catch {
    // Linter infrastructure failed — continue without linter output
  }

  const prompt = buildReviewPrompt(
    agent,
    fileContent,
    input.file_path,
    input.context,
    linterOutput
  );

  return { content: prompt, isError: false };
}

File: rust/src/tools/review.rs
use crate::types::AgentDefinition;
use crate::tools::linter;

pub use super::ToolResult;

/// Language hint from file extension.
fn lang_for_ext(ext: &str) -> &str {
    match ext {
        "ts" | "tsx" => "typescript",
        "js" | "jsx" => "javascript",
        "py" => "python",
        "rs" => "rust",
        "go" => "go",
        "java" => "java",
        "php" => "php",
        "cs" => "csharp",
        "c" | "h" => "c",
        "dart" => "dart",
        "lua" => "lua",
        "sql" => "sql",
        "proto" => "protobuf",
        _ => "",
    }
}

fn build_review_prompt(
    agent: &AgentDefinition,
    file_content: &str,
    file_path: &str,
    context: Option<&str>,
    linter_output: Option<&str>,
) -> String {
    let mut parts = Vec::new();

    parts.push(agent.system_prompt.clone());

    if !agent.checklist.is_empty() {
        parts.push("\n## Review Checklist\n".to_string());
        parts.push(agent.checklist.clone());
    }

    parts.push("\n## File Under Review\n".to_string());
    parts.push(format!("**Path:** `{file_path}`\n"));

    let ext = file_path.rsplit('.').next().unwrap_or("");
    let lang = lang_for_ext(ext);
    parts.push(format!("```{lang}"));
    parts.push(file_content.to_string());
    parts.push("```".to_string());

    if let Some(lo) = linter_output {
        if !lo.is_empty() {
            parts.push(lo.to_string());
        }
    }

    if let Some(ctx) = context {
        parts.push("\n## Additional Context\n".to_string());
        parts.push(ctx.to_string());
    }

    // Linter cross-reference note
    if linter_output.is_some_and(|lo| !lo.is_empty()) {
        parts.push("\n## Note\n".to_string());
        parts.push("If linter findings are provided above, reference them where relevant — confirm, expand on, or contextualize what the tools found.".to_string());
    }

    parts.join("\n")
}

const MAX_FILE_SIZE: usize = 512 * 1024; // 512KB

pub async fn execute_review(
    agent: &AgentDefinition,
    file_path: &str,
    context: Option<&str>,
) -> ToolResult {
    // Read file
    let file_content = match tokio::fs::read_to_string(file_path).await {
        Ok(c) => c,
        Err(e) => {
            return ToolResult {
                content: format!(
                    "Cannot read file: {file_path} — {e}. Check that the path is absolute and the file exists."
                ),
                is_error: true,
            };
        }
    };

    if file_content.len() > MAX_FILE_SIZE {
        let size_kb = file_content.len() / 1024;
        return ToolResult {
            content: format!(
                "File skipped: {file_path} is {size_kb}KB (limit: {}KB). Likely generated or minified — exclude from review.",
                MAX_FILE_SIZE / 1024
            ),
            is_error: false,
        };
    }

    // Run linters (best-effort)
    let linter_output = match linter::run_linters(file_path, &agent.recommended_tools).await {
        Ok(result) => {
            let formatted = linter::format_linter_findings(&result);
            if formatted.is_empty() { None } else { Some(formatted) }
        }
        Err(_) => None,
    };

    let prompt = build_review_prompt(
        agent,
        &file_content,
        file_path,
        context,
        linter_output.as_deref(),
    );

    ToolResult {
        content: prompt,
        is_error: false,
    }
}
